lib/Form/Processor/RoleFormProcessor.php

```diff
@@ -37,28 +37,62 @@ public static function getSubscribedEvents()
 
     public function processDefaultAction(FormActionEvent $event)
     {
-        // Don't update anything if we just want to cancel the draft.
-        if ($event->getClickedButton() === 'removeDraft') {
+        if (self::skipDefaultSave($event)) {
             return;
         }
 
-        /** @var \EzSystems\RepositoryForms\Data\Role\RoleData $roleData */
-        $roleData = $event->getData();
-        $roleDraft = $roleData->roleDraft;
-        $this->roleService->updateRoleDraft($roleDraft, $roleData);
+        $this->processSaveRole($event);
     }
 
     public function processSaveRole(FormActionEvent $event)
     {
-        /** @var RoleDraft $roleDraft */
-        $roleDraft = $event->getData()->roleDraft;
+        $roleDraft = self::getRoleDraft($event);
+
+        $this->roleService->updateRoleDraft($roleDraft, self::getRoleData($event));
         $this->roleService->publishRoleDraft($roleDraft);
     }
 
     public function processRemoveDraft(FormActionEvent $event)
     {
-        /** @var RoleDraft $roleDraft */
-        $roleDraft = $event->getData()->roleDraft;
-        $this->roleService->deleteRoleDraft($roleDraft);
+        $this->roleService->deleteRoleDraft(self::getRoleDraft($event));
+    }
+
+    /**
+     * Returns true if saving should not be done in the default process.
+     *
+     * There's no need to save when cancelling the draft, nor when the Save button is clicked, as the save process
+     * takes care of it. Saving is only needed when the form has been submitted by pressing Enter.
+     *
+     * @param FormActionEvent $event
+     *
+     * @return bool
+     */
+    protected static function skipDefaultSave(FormActionEvent $event)
```

Comment:
@bdunogier Using `static` because I can. Is the best practice to only do it when it actually avoids instantiation?

"skip" was your proposal, man :) But I can change it.